### PR TITLE
8255466: C2 crashes at ciObject::get_oop() const+0x0

### PR DESCRIPTION
**Pull request by vnkozlov**
Graal testing hit this issue with product VM. Tom R. suggested that it could be the case of reflective unsafe static field access that would eventually be optimized away because the Class is null:
`if (staticFieldBase != null) {
  return Unsafe.getInt(staticFieldBase, Unsafe.staticFieldOffset(field));
}`

I suggest to replace assert with runtime check. Note, `o` value is assigned to `_const_oop` so semantically new code is the same except additional runtime check.

I also noticed that const_oop is accessed without check for NULL in new Vector API code. I added check there too.

Passed tier1-3 testing.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8255466](https://bugs.openjdk.java.net/browse/JDK-8255466): C2 crashes at ciObject::get_oop() const+0x0

### Reviewers
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to da8be5298c8d22bf8e1b26f3d943c26091df5f76

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/890/head:pull/890`
`$ git checkout pull/890`

### CODE REVIEW

File: src/hotspot/share/opto/type.cpp [hunk 1]
```diff
@@ -3046,9 +3046,11 @@ TypeOopPtr::TypeOopPtr(TYPES t, PTR ptr, ciKlass* k, bool xk, ciObject* o, int o
         } else if (klass() == ciEnv::current()->Class_klass() &&
                    _offset >= InstanceMirrorKlass::offset_of_static_fields()) {
```
Comment (iwanowww) [inline src/hotspot/share/opto/type.cpp:3047]: You could turn the assert into the check in the enclosing `if`. IMO it makes the code clearer.

**Review by iwanowww — APPROVED** [review]
It would be nice to have a regression test for it.

Otherwise, looks good.

File: src/hotspot/share/opto/type.cpp (hunk 1 above)
Comment (vnkozlov) [reply src/hotspot/share/opto/type.cpp:3047]: In that case it would fall into `Instance fields` code below under `} else {` which I don't want.

**Pull request closed without merging**